d2/src/main/java/com/linkedin/d2/xds/XdsToD2PropertiesAdaptor.java

```diff
@@ -533,67 +541,181 @@ public void onReconnect()
     {
       notifyAvailabilityChanges(true);
     }
+  }
+
+  private class UriPropertiesCollectionResourceWatcher implements XdsClient.D2URICollectionResourceWatcher
+  {
+    final String _clusterName;
+    final AtomicBoolean _isInit;
+    final long _initFetchStart;
+
+    Map<String, UriProperties> _currentData = new HashMap<>();
+
+    public UriPropertiesCollectionResourceWatcher(String clusterName)
+    {
+      _clusterName = clusterName;
+      _isInit = new AtomicBoolean(true);
+      _initFetchStart = System.nanoTime();
+    }
 
-    private void emitSDStatusInitialRequestEvent(String cluster, boolean succeeded)
+    @Override
+    public void onChanged(XdsClient.D2URICollectionUpdate update)
     {
-      if (_eventEmitter == null)
+      boolean isInit = _isInit.compareAndSet(true, false);
+      if (isInit)
       {
-        LOG.info("Service discovery event emitter in XdsToD2PropertiesAdaptor is null. Skipping emitting events.");
-        return;
+        emitSDStatusInitialRequestEvent(_clusterName, true, _initFetchStart);
       }
 
-      // measure request duration and convert to milli-seconds
-      long initialFetchDurationMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - _initFetchStart);
-      if (initialFetchDurationMillis < 0)
+      if (_uriEventBus != null)
       {
-        LOG.warn("Failed to log ServiceDiscoveryStatusInitialRequest event, initialFetchStartAt time is greater than current time.");
-        return;
+        try
+        {
+          Map<String, UriProperties> newData = new HashMap<>(_currentData);
+          for (String removedResource : update.getRemovedUris())
+          {
+            newData.remove(removedResource);
+          }
+          for (Map.Entry<String, XdsD2.D2URI> entry : update.getUris().entrySet())
+          {
+            XdsD2.D2URI d2URI = entry.getValue();
+            UriProperties uriProperties = _uriPropertiesJsonSerializer.fromProto(d2URI);
+            if (uriProperties.getVersion() < 0)
+            {
+              LOG.warn("xDS data for {} has invalid version: {}", entry.getKey(), uriProperties.getVersion());
+            }
+            newData.put(entry.getKey(), uriProperties);
+          }
```

Comment:
these has been changed in 29.51.2. Please rebase the latest code.

Reply:
This is on top of https://github.com/linkedin/rest.li/pull/975, so I will wait for @brycezhongqing to rebase